### PR TITLE
Finalize public formats API and add test cases

### PR DESCRIPTION
**Pull request by amorde**
Updates the IPv4 and IPv6 implementations and adds test cases for IPv4, IPv6, and a custom added format.

This includes the changes from #14 and #17

Note: I wrapped the validator passed to `addFormat` in a `string` type validator, as the formats functionality only makes sense for strings.

### CODE REVIEW

**Comment by amorde** [comment]
@kylef Any way you could take a look at this? It would be great if these updates could be includes & pushed to CocoaPods

File: README.md [hunk 1]
```diff
@@ -27,6 +33,25 @@ let schema = Schema([
...
+// Note: custom formats will automatically be combined with a string type validator
+schema.addFormat(formatKey: "alpha") { (value) -> (ValidationResult) in
+    if let str = value as? String {
+        let allowedCharacters = CharacterSet.letters
+        let c = str.components(separatedBy: allowedCharacters)
+        let leftover = (c as NSArray).componentsJoined(by: "")
+        if leftover.characters.count > 0 {
+            return .invalid(["\(str) contains non-alpha characters"])
+        }
+        return .Valid
```
Comment (kylef) [inline README.md]: Should this not be `.valid` in Swift 3?

Comment (amorde) [inline README.md]: I had the same thought myself, I'll go ahead and make that change

**Comment by amorde** [comment]
@kylef Implemented that change and added it to the changelog

**Comment by amorde** [comment]
@kylef Just checking in, any chance of getting this merged?

**Comment by kylef** [comment]
I apologise that I never made time to reply here. After giving this some more though, I do not think it makes sense for the formats to be registered on the schema itself. The Schema should just be a schema and not contain logic such as formats. I think we should create separate a `Validator` class which contains these formats and provides a validate method taking the schema and the value.

If you are interested in refactoring this, that would be great. Otherwise I will try and find some time to extend this PR refactoring out validation to a separate `Validator`.

**Comment by kylef** [comment]
I appreciate the effort here so thank you. I do have one suggestion for the future, this PR also does three separate change, I'd recommend in the future making three separate PRs which make each individual change, it will be easier to review and easier to integrate into master.

- Swift 3 support
- Refactoring IPv4/IPv6 format
- Supporting multiple formats.

**Comment by amorde** [comment]
Thanks for the feedback! I think I'll go ahead and close this, and maybe factor out some of the changes into another PR

**Comment by kylef** [comment]
Spent some time refactoring the library, validations have been broken into a new set of classes, `Draft4Validator`, `Draft6Validator` and `Draft7Validator` which all contain (at the moment, private) `formats`.

Usage is described in https://github.com/kylef/JSONSchema.swift/issues/24#issuecomment-526803738.

**Pull request closed without merging**